cache/pmdk_cache.h

```diff
@@ -0,0 +1,345 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+//
+// Copyright (c) 2011 The LevelDB Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+#pragma once
+
+#include <string>
+
+#include "cache/sharded_cache.h"
+
+#include "port/malloc.h"
+#include "port/port.h"
+#include "util/autovector.h"
+
+// TODO: re-order included headers.
+#include <unistd.h>
+#include <libpmemobj++/p.hpp>
+#include <libpmemobj++/container/concurrent_hash_map.hpp>
+#include <libpmemobj++/pool.hpp>
+#include <libpmemobj++/persistent_ptr.hpp>
+#include <libpmemobj++/make_persistent_array.hpp>
+#include <libpmemobj++/transaction.hpp>
+
+namespace po = pmem::obj;
+
+namespace ROCKSDB_NAMESPACE {
+
+struct HandleClassifier{
+  void* foo;
+  void* type;
+};
+
+struct PersistentEntry;
+
+// This will be the transient handle of the cache.
+struct TransientHandle {
+  void* value;
+  // a fixed number to tell TransientHandle and LRUHandle apart.
+  void* type = reinterpret_cast<void*>(0x1);
```

Comment:
iiuc, `TransientHandle::type` and `HandleClassifier::type` have to be of the same type. Also, `HandleClassifier::foo` and `TransientHandle::value` must be of the same type. Does `type` have to be a `void*`?

Reply:
> Does `type` have to be a `void*`?

No. It's just a convenient way of having `foo` and `type` the same _size_ as the first two fields of both `TransientHandle` and `LRUCache::LRUHandle`. Please see `IsLRUHandle()` for the usage of this. It uses the fact that the second field of `LRUHandle` is a pointer (null or some address), but the second field of `TransientHandle` is `0x1`.

But overall, this is not at all a general way to tell apart `TransientHandle` and `LRUHandle`.